src/graphics/text.rs

```diff
@@ -70,44 +84,96 @@ impl Text {
         Text {
             content,
             font,
-            size: Scale::uniform(size),
+            scale: Scale::uniform(size),
+            size: RefCell::new(TextSize::Unknown),
             quads: RefCell::new(Vec::new()),
         }
     }
 
     /// Sets the content of the text.
     pub fn set_content<S: Into<String>>(&mut self, content: S) {
         self.content = content.into();
+        self.invalidate();
+    }
+
+    /// Get the outer bounds of the text when rendered to the screen.
+    /// If the text is not rendered yet, this method will re-render it and calculate the bounds.
+    /// The size is automatically cached, so calling this multiple times will only render once.
+    pub fn get_size(&self, ctx: &mut Context) -> Option<graphics::Rectangle> {
+        match *self.size.borrow() {
+            TextSize::Unknown => {}
+            TextSize::NoSize => return None,
+            TextSize::Known(r) => return Some(r),
+        }
+        self.attempt_recalculate_quads(ctx);
+        match *self.size.borrow() {
+            TextSize::Unknown => unreachable!(),
+            TextSize::NoSize => None,
+            TextSize::Known(r) => Some(r),
+        }
     }
 
     /// Sets the font of the text.
     pub fn set_font(&mut self, font: Font) {
         self.font = font;
-        self.quads = RefCell::new(Vec::new());
+        self.invalidate();
     }
 
     /// Sets the size of the text.
-    pub fn set_size(&mut self, size: f32) {
-        self.size = Scale::uniform(size);
-        self.quads = RefCell::new(Vec::new());
+    pub fn set_scale(&mut self, size: f32) {
+        self.scale = Scale::uniform(size);
+        self.invalidate();
     }
-}
 
-impl Drawable for Text {
-    fn draw<T: Into<DrawParams>>(&self, ctx: &mut Context, params: T) {
-        let params = params.into();
-        let transform = params.build_matrix();
+    /// Invalidates this Text component, forcing it to redraw itself on the next draw pass.
+    fn invalidate(&mut self) {
+        self.quads = RefCell::new(Vec::new());
+        self.size = RefCell::new(TextSize::Unknown);
+    }
 
+    /// Attempt to recalculate the fontquads
+    fn attempt_recalculate_quads(&self, ctx: &mut Context) {
         let section = Section {
             text: &self.content,
-            scale: self.size,
+            scale: self.scale,
             font_id: self.font.id,
             ..Section::default()
         };
 
         if let BrushAction::Draw(new_quads) = draw_text(ctx, section) {
+            if new_quads.is_empty() {
+                *self.size.borrow_mut() = TextSize::NoSize;
+            } else {
+                let mut max_x = std::f32::MIN;
+                let mut max_y = std::f32::MIN;
+                let mut min_x = std::f32::MAX;
+                let mut min_y = std::f32::MAX;
+                for quad in &new_quads {
+                    max_x = max_x.max(quad.x1).max(quad.x2);
```

Comment:
Would `quad.x1.max(quad.x2)` not suffice? Or is there some edge case where this is neccecary?

Reply:
Is there a difference between `max_x.max(quad.x1).max(quad.x2)` and `max_x.max(quad.x1.max(quad.x2))`? We definitely have to keep track of `max_x` because we get multiple quads back from the `draw_text`

Reply:
I'm not sure if `draw_text` has any guarantees about which order the quads come back. If they're sorted properly, we could just get `min_` from the first quad and `max_` from the last one, but I didn't want to take that risk.

Reply:
Aaah, I get it, it's just a default value for the first iteration of the loop. Sorry, me being slow :)